apps/pipelines/models.py

```diff
@@ -127,22 +163,54 @@
             session=session,
         )
 
-    def _save_message_to_history(self, session: ExperimentSession, message: str, type_: ChatMessageType):
-        ChatMessage.objects.create(
+    def _save_message_to_history(self, session: ExperimentSession, message: str, type_: ChatMessageType) -> ChatMessage:
+        return ChatMessage.objects.create(
             chat=session.chat,
             message_type=type_.value,
             content=message,
         )
         # TODO: Add tags here?
 
+    @transaction.atomic()
+    def create_new_version(self, *args, **kwargs):
+        version_number = self.version_number
+        self.version_number = version_number + 1
+        self.save()
+        pipeline_version = super().create_new_version(*args, **kwargs)
+        pipeline_version.version_number = version_number
+        pipeline_version.save(update_fields=["version_number"])
 
-class Node(BaseModel):
+        new_nodes = []
+        for node in self.node_set.all():
+            node_version = node.create_new_version()
+            node_version.pipeline = pipeline_version
+            new_nodes.append(node_version)
+
+        Node.objects.bulk_update(new_nodes, fields=["pipeline"])
+
+        return pipeline_version
+
+    @transaction.atomic()
+    def archive(self):
+        super().archive()
+        self.node_set.update(is_archived=True)
+
+
+class Node(BaseModel, VersionsMixin):
     flow_id = models.CharField(max_length=128, db_index=True)  # The ID assigned by react-flow
     type = models.CharField(max_length=128)  # The node type, should be one from nodes/nodes.py
     label = models.CharField(max_length=128, blank=True, default="")  # The human readable label
     params = models.JSONField(default=dict)  # Parameters for the specific node type
-
+    working_version = models.ForeignKey(
+        "self",
+        on_delete=models.CASCADE,
+        null=True,
+        blank=True,
+        related_name="versions",
+    )
```

Comment:
This will lead to an unwanted DB state. We always want to keep the reference to the `working_version`. I'm using the `is_archived` to archive the versions. We never delete Nodes, but only archive them so that we can keep the reference to the working version in tact. While you are here, would you please look for any places where we actually delete a node instead of archiving? (Obviously excluding tests)

Reply:
I see. In this case it's fine for it to be deleted instead of archived. Thanks

Reply:
yes please